src/CodeGen_C.cpp

```diff
@@ -1489,6 +1489,252 @@ void CodeGen_C::forward_declare_type_if_needed(const Type &t) {
     forward_declared.insert(t.handle_type);
 }
 
+void CodeGen_C::emit_argv_wrapper(const std::string &function_name,
+                                  const std::vector<LoweredArgument> &args) {
+    if (is_header_or_extern_decl()) {
+        stream << "\nHALIDE_FUNCTION_ATTRS\nint " << function_name << "_argv(void **args);\n";
+        return;
+    }
+
+    stream << "\nHALIDE_FUNCTION_ATTRS\nint " << function_name << "_argv(void **args) {\n";
+    indent += 1;
+
+    stream << get_indent() << "return " << function_name << "(\n";
+    indent += 1;
+
+    for (size_t i = 0; i < args.size(); i++) {
+        if (args[i].is_buffer()) {
+            stream << get_indent() << "(halide_buffer_t *)args[" << i << "]";
+        } else if (args[i].name == "__user_context") {
+            stream << get_indent() << "args[" << i << "]";
+        } else {
+            stream << get_indent() << "*(" << type_to_c_type(args[i].type, false) << " const *)args[" << i << "]";
+        }
+        if (i + 1 < args.size()) {
+            stream << ",";
+        }
+        stream << "\n";
+    }
+
+    indent -= 1;
+    stream << ");\n";
+
+    indent -= 1;
+    stream << "}";
+}
+
+void CodeGen_C::emit_metadata_getter(const std::string &function_name,
+                                     const std::vector<LoweredArgument> &args,
+                                     const std::map<std::string, std::string> &metadata_name_map) {
+    if (is_header_or_extern_decl()) {
+        stream << "\nHALIDE_FUNCTION_ATTRS\nconst struct halide_filter_metadata_t *" << function_name << "_metadata();\n";
+        return;
+    }
+
+    auto map_name = [&metadata_name_map](const std::string &from) -> std::string {
+        auto it = metadata_name_map.find(from);
+        return it == metadata_name_map.end() ? from : it->second;
+    };
+
+    auto legalize_name = [](const std::string &from) -> std::string {
```

Comment:
There is a `c_print_name` function, which does something similar. Is it possible to use it here instead?

Reply:
Done